grouped-fields: Add new Grouped fields component. Update inputs to support it (#1566)
agriculturegovau/agds-next: packages/react/src/grouped-fields/docs/overview.mdx

---
title: Grouped fields
description: The Grouped fields component allows two closely related form inputs to be grouped together.
group: Forms
storybookPath: /story/forms-groupedfields--basic
unreleased: true

> figmaGalleryNodeId: REPLACE_ME

Author: TODO when it's added to Figma

relatedComponents:
  [
    'autocomplete',
    'combobox',
    'date-picker',
    'password-input',
    'search-input',
    'select',
    'text-input',
    'textarea',
  ]
---

```jsx live
<GroupedFields legend="Item weight">
	{({ field1Props, field2Props }) => (
		<>
			<TextInput label="Weight" {...field1Props} />
			<Select
				label="Unit"
				options={[
					{ value: 'mg', label: 'mg' },
					{ value: 'g', label: 'g' },
					{ value: 'kg', label: 'kg' },
					{ value: 'ton', label: 'ton' },
				]}
				hideOptionalLabel
				maxWidth="sm"
				value="kg"
				{...field2Props}
			/>
		</>
	)}
</GroupedFields>
```

Grouped fields allows two related form inputs to be composed together while providing the ability to present and describe this relationship and any invalid input errors.

It uses the [render props pattern](https://www.patterns.dev/react/render-props-pattern/) to spread a series of necessary attributes to the `children` form inputs.

<DoHeading />

- group two related inputs, such as weight and unit, and date and time

> - always provide a meaningful legend, regardless if it’s visually hidden

Reviewer: ```suggestion
- always provide a meaningful legend, if it is visible or visually hidden
```

Author: I feel like we need "regardless" again:

- always provide a meaningful legend, regardless if it is visible or visually hidden

Reviewer: agreed, please proceed

- use the group’s optional label instead of the inputs’ labels when both fields are optional.
- provide hint text for instruction

<DontHeading />

> - group unrelated inputs as a space saving exercise.

Reviewer: ```suggestion
- group unrelated inputs to save space or reduce the length of a form
```

> ## Hide Optional Label

Reviewer: ```suggestion
## Hide optional label
```

When both inputs are required or one input is optional and the other has no empty state, hide the group’s optional label with the `hideOptionalLabel` prop.

```jsx live
<FormStack>
	<GroupedFields legend="Item weight" hideOptionalLabel>
		{({ field1Props, field2Props }) => (
			<>
				<TextInput label="Weight" {...field1Props} required />
				<Select
					label="Unit"
					options={[
						{ value: 'mg', label: 'mg' },
						{ value: 'g', label: 'g' },
						{ value: 'kg', label: 'kg' },
						{ value: 'ton', label: 'ton' },
					]}
					maxWidth="sm"
					placeholder="Choose"
					required
					{...field2Props}
				/>
			</>
		)}
	</GroupedFields>

	<GroupedFields legend="Item weight" hideOptionalLabel>
		{({ field1Props, field2Props }) => (
			<>
				<TextInput label="Weight" {...field1Props} />
				<Select
					label="Unit"
					options={[
						{ value: 'mg', label: 'mg' },
						{ value: 'g', label: 'g' },
						{ value: 'kg', label: 'kg' },
						{ value: 'ton', label: 'ton' },
					]}
					maxWidth="sm"
					value="kg"
					hideOptionalLabel
					{...field2Props}
				/>
			</>
		)}
	</GroupedFields>
</FormStack>
```

> ## Visually Hide Legend

Reviewer: ```suggestion
## Visually hide legend
```

				]}
				maxWidth="sm"
				placeholder="Choose"
				required
				{...field2Props}
			/>
		</>
	)}
</GroupedFields>
```

## Hint

Use the `hint` prop to provide help that’s relevant to the majority of users, like how their information will be used, or where to find it.

Don’t use long paragraphs and lists in hint text. Screen readers read out the entire text when users interact with the form element. This could frustrate users if the text is long.

Don’t include links within hint text. While screen readers will read out the link text when describing the field, they will not tell users that the text is a link.

```jsx live
<GroupedFields legend="Item weight" hint="Example hint text">
	{({ field1Props, field2Props }) => (
		<>
			<TextInput label="Weight" {...field1Props} required />
			<Select
				label="Unit"
				options={[
					{ value: 'mg', label: 'mg' },
					{ value: 'g', label: 'g' },
					{ value: 'kg', label: 'kg' },
					{ value: 'ton', label: 'ton' },
				]}
				maxWidth="sm"
				value="kg"
				{...field2Props}
			/>
		</>
	)}
</GroupedFields>
```

## Invalid

Use the `field1Invalid`, `field2Invalid` and `message` props to indicate if the user input is invalid.

> The `message` should be generic so it can apply and make sense to both inputs.

Reviewer: ```suggestion
The `message` should be helpful but generic enough to apply and make sense to both inputs.
```
